### ISSUE TITLE
neovim update to 0.10.0

### ISSUE THREAD

**Issue by hjm7** [msys2/MINGW-packages#21313]
### Package name

mingw-w64-clang-x86_64-neovim

### New version number

 0.10.0

### Other information that may be useful

Nvim 0.10 introduces a new 'termsync' option (enabled by default) which enables [synchronized output](https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036) feature if it is supported by the terminal emulator.

**Comment by prgreadonly** [comment]
Would be nice to see this for ucrt64 as well

**Comment by sewbacca** [comment]
I gave updating neovim to v0.10.0 a go, since lazydev requires this version, but didn't finish it yet. I got halfway through though, so if anyone wants to give it a try, here is what I've found out so far:
- The patch `0001-cmake-disable-bundling.patch` needs to be updated to:
```diff
diff --git a/src/nvim/CMakeLists.txt b/src/nvim/CMakeLists.txt
index d9cc695c..42465eca 100644
--- a/src/nvim/CMakeLists.txt
+++ b/src/nvim/CMakeLists.txt
@@ -726,7 +726,7 @@ if(ENABLE_LTO)
 endif()

 add_custom_target(nvim_runtime_deps)
-if(WIN32)
+if(FALSE)
   # Copy DLLs and third-party tools to bin/ and install them along with nvim
   add_custom_command(TARGET nvim_runtime_deps
     COMMAND ${CMAKE_COMMAND} -E ${COPY_DIRECTORY} ${PROJECT_BINARY_DIR}/windows_runtime_deps/
```
- Add `0002-cmake-fix-FindLpeg.patch` to fix finding `lpeg` both during configuration and compilation
```diff
diff --git a/cmake/FindLpeg.cmake b/cmake/FindLpeg.cmake
index 3d0ff5929..7e6bca45a 100644
--- a/cmake/FindLpeg.cmake
+++ b/cmake/FindLpeg.cmake
@@ -1,3 +1,4 @@
+set(CMAKE_FIND_LIBRARY_SUFFIXES .dll ${CMAKE_FIND_LIBRARY_SUFFIXES})
 find_library2(LPEG_LIBRARY NAMES lpeg_a lpeg liblpeg_a lpeg${CMAKE_SHARED_LIBRARY_SUFFIX} PATH_SUFFIXES lua/5.1)

 find_package_handle_standard_args(Lpeg DEFAULT_MSG LPEG_LIBRARY)
diff --git a/src/nvim/generators/preload.lua b/src/nvim/generators/preload.lua
index e1467107..773483f6 100644
--- a/src/nvim/generators/preload.lua
+++ b/src/nvim/generators/preload.lua
@@ -6,6 +6,8 @@ package.path = gendir .. '/?.lua;' .. package.path
 _G.vim = require 'vim.shared'
 _G.vim.inspect = require 'vim.inspect'
 package.cpath = package.cpath .. ';' .. nlualib
+-- Force loading lpeg first
+package.loadlib(package.searchpath("lpeg", package.cpath), "*")
 require 'nlua0'
 vim.NIL = vim.mpack.NIL -- WOW BOB WOW
 ```
The lpeg library causes some problems however: Since it is a lua module, there is no need for an importlib, which CMake seems to expect (see #6394). I adjusted `FindLpeg.cmake` with the workaround provided by the issue. However when running the `preload.lua` during generation, `libnlua0.dll` links to lpeg, but cannot find the library (since it is in `lib/lua/5.1`), which fails to load the lib. I use `package.loadlib` to link `lpeg` during runtime (in a LuaJIT/Lua5.2+ way) before `libnlua0.dll` gets loaded, which gets me halfway through the actual compilation process, until another error occurs, which I don't have the time to figure out right now.

**Comment by aitorciki** [comment]
I've given it a go, following @sewbacca steps, with no luck 😞 

`lpeg` is properly found during configuration using @sewbacca's patch, or also explicitly using `-DLPEG_LIBRARY` like [Homebrew ](https://github.com/Homebrew/homebrew-core/blob/e68e8c162d12a4227b5ceb3104eebb3b1fc1ecbd/Formula/n/neovim.rb#L121)does, but it's not properly linked afterwards.

It's properly found:

```
-- Found Lpeg: C:/msys64/mingw64/lib/lua/5.1/lpeg.dll
```

It's also referenced when building `libnlua0.dll`:

<pre>
[8/419] C:\WINDOWS\system32\cmd.exe /C
"cd . && C:\msys64\mingw64\bin\gcc.exe -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 
-fstack-protector-strong -Wno-format -Wno-format-extra-args -O3 -DNDEBUG   -shared <b>-o lib\libnlua0.dll</b> 
-Wl,--major-image-version,0,--minor-image-version,0 src/nvim/CMakeFiles/nlua0.dir/__/mpack/conv.c.obj 
src/nvim/CMakeFiles/nlua0.dir/__/mpack/lmpack.c.obj src/nvim/CMakeFiles/nlua0.dir/__/mpack/mpack_core.c.obj 
src/nvim/CMakeFiles/nlua0.dir/__/mpack/object.c.obj src/nvim/CMakeFiles/nlua0.dir/__/mpack/rpc.c.obj 
src/nvim/CMakeFiles/nlua0.dir/__/nlua0.c.obj  <b>C:/msys64/mingw64/lib/lua/5.1/lpeg.dll</b>  
C:/msys64/mingw64/lib/libluajit-5.1.dll.a  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 
-luuid -lcomdlg32 -ladvapi32 && cd ."
</pre>

But the resulting `libnlua0.dll` isn't properly linked to `liblpeg`:

<pre>
$ ldd src/build-MINGW64/lib/libnlua0.dll
        ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ffed3110000)
        KERNEL32.DLL => /c/WINDOWS/System32/KERNEL32.DLL (0x7ffed1a90000)
        KERNELBASE.dll => /c/WINDOWS/System32/KERNELBASE.dll (0x7ffed0790000)
        msvcrt.dll => /c/WINDOWS/System32/msvcrt.dll (0x7ffed1b80000)
        libnlua0.dll => /home/aitor/src/mingw-w64-neovim-0.10/src/build-MINGW64/lib/libnlua0.dll (0x7ffe9c1e0000)
        lua51.dll => /mingw64/bin/lua51.dll (0x7ffe9c140000)
        <b>lpeg.dll => not found</b>
</pre>

Nor is `nvim.exe` afterwards, which results in a broken binary and the build failing when that binary is invoked to generate other artifacts:

<pre>
$ ldd src/build-MINGW64/bin/nvim.exe
        ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ffed3110000)
        KERNEL32.DLL => /c/WINDOWS/System32/KERNEL32.DLL (0x7ffed1a90000)
        KERNELBASE.dll => /c/WINDOWS/System32/KERNELBASE.dll (0x7ffed0790000)
        msvcrt.dll => /c/WINDOWS/System32/msvcrt.dll (0x7ffed1b80000)
        ole32.dll => /c/WINDOWS/System32/ole32.dll (0x7ffed0e50000)
        msvcp_win.dll => /c/WINDOWS/System32/msvcp_win.dll (0x7ffed0b40000)
        NETAPI32.dll => /c/WINDOWS/SYSTEM32/NETAPI32.dll (0x7ffebd390000)
        ucrtbase.dll => /c/WINDOWS/System32/ucrtbase.dll (0x7ffed0c10000)
        GDI32.dll => /c/WINDOWS/System32/GDI32.dll (0x7ffed1210000)
        win32u.dll => /c/WINDOWS/System32/win32u.dll (0x7ffed0760000)
        gdi32full.dll => /c/WINDOWS/System32/gdi32full.dll (0x7ffed0d30000)
        USER32.dll => /c/WINDOWS/System32/USER32.dll (0x7ffed1c30000)
        combase.dll => /c/WINDOWS/System32/combase.dll (0x7ffed2d40000)
        RPCRT4.dll => /c/WINDOWS/System32/RPCRT4.dll (0x7ffed1370000)
        WS2_32.dll => /c/WINDOWS/System32/WS2_32.dll (0x7ffed1ed0000)
        libiconv-2.dll => /mingw64/bin/libiconv-2.dll (0x7ffe9c110000)
        libluv.dll => /mingw64/bin/libluv.dll (0x7ffe9c0d0000)
        libintl-8.dll => /mingw64/bin/libintl-8.dll (0x7ffe98ff0000)
        lua51.dll => /mingw64/bin/lua51.dll (0x7ffe84730000)
        ADVAPI32.dll => /c/WINDOWS/System32/ADVAPI32.dll (0x7ffed12b0000)
        sechost.dll => /c/WINDOWS/System32/sechost.dll (0x7ffed1f50000)
        bcrypt.dll => /c/WINDOWS/System32/bcrypt.dll (0x7ffed0be0000)
        libunibilium-4.dll => /mingw64/bin/libunibilium-4.dll (0x7ffecb210000)
        libuv-1.dll => /mingw64/bin/libuv-1.dll (0x7ffe3a3e0000)
        libvterm-0.dll => /mingw64/bin/libvterm-0.dll (0x7ffeacfa0000)
        SHELL32.dll => /c/WINDOWS/System32/SHELL32.dll (0x7ffed2110000)
        libmsgpackc.dll => /mingw64/bin/libmsgpackc.dll (0x7ffeacbf0000)
        libtree-sitter.dll => /mingw64/bin/libtree-sitter.dll (0x7ffe987c0000)
        dbghelp.dll => /c/WINDOWS/SYSTEM32/dbghelp.dll (0x7ffec6df0000)
        IPHLPAPI.DLL => /c/WINDOWS/SYSTEM32/IPHLPAPI.DLL (0x7ffeceef0000)
        OLEAUT32.dll => /c/WINDOWS/System32/OLEAUT32.dll (0x7ffed1530000)
        USERENV.dll => /c/WINDOWS/SYSTEM32/USERENV.dll (0x7ffecfad0000)
        <b>lpeg.dll => not found</b>
</pre>

Checking other distributions, this library is properly linked. Here is ArchLinux for example:

<pre>
$ ldd /usr/bin/nvim
	libluv.so.1 => /usr/lib/libluv.so.1 (0x00007fffff1ea000)
	libvterm.so.0 => /usr/lib/libvterm.so.0 (0x00007fffff1d5000)
	<b>/usr/lib/lua/5.1/lpeg.so (0x00007fffff1c5000)</b>
	libmsgpackc.so.2 => /usr/lib/libmsgpackc.so.2 (0x00007fffff1bc000)
	libtree-sitter.so.0 => /usr/lib/libtree-sitter.so.0 (0x00007fffff18e000)
	libunibilium.so.4 => /usr/lib/libunibilium.so.4 (0x00007fffff17f000)
	libluajit-5.1.so.2 => /usr/lib/libluajit-5.1.so.2 (0x00007fffff0ef000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007fffff002000)
	libuv.so.1 => /usr/lib/libuv.so.1 (0x00007ffffefce000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007ffffefa1000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007ffffedb5000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007ffffffca000)
</pre>

Or even MacOS's Homebrew binary:

<pre>
otool -L /opt/homebrew/Cellar/neovim/0.10.1/bin/nvim
/opt/homebrew/Cellar/neovim/0.10.1/bin/nvim:
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1226.0.0)
	/opt/homebrew/opt/luv/lib/libluv.1.dylib (compatibility version 1.0.0, current version 1.48.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/opt/homebrew/opt/libvterm/lib/libvterm.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	<b>/opt/homebrew/opt/lpeg/lib/lua/5.1/lpeg.so (compatibility version 0.0.0, current version 0.0.0)</b>
	/opt/homebrew/opt/msgpack/lib/libmsgpack-c.2.dylib (compatibility version 2.0.0, current version 2.0.0)
	/opt/homebrew/opt/tree-sitter/lib/libtree-sitter.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/unibilium/lib/libunibilium.4.dylib (compatibility version 5.0.0, current version 5.1.0)
	/opt/homebrew/opt/gettext/lib/libintl.8.dylib (compatibility version 13.0.0, current version 13.0.0)
	/opt/homebrew/opt/luajit/lib/libluajit-5.1.2.dylib (compatibility version 2.1.0, current version 2.1.255)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.100.2)
	/usr/lib/libutil.dylib (compatibility version 1.0.0, current version 1.0.0)
	/opt/homebrew/opt/libuv/lib/libuv.1.dylib (compatibility version 2.0.0, current version 2.0.0)
</pre>

My understanding of C compilers & linkers is rather basic, so I don't know where to go from here, since these other distributions don't seem to do anything special for that library to be linked properly during the build.

**Comment by sewbacca** [comment]
@aitorciki thanks for giving it a go. As of my understanding, the issue is that for normal distributions, the dynamic linker is using `LDPATH` which contains the `lib` directory to find dynamically linked libraries. Windows doesn't have a `LDPATH`, it uses `PATH` to find the necessary libraries. This is mostly fine, since all libraries for MINGW are installed to the `bin` directory instead of the `lib` directory per default. Lua modules (and lpeg is a Lua module) are not installed to the bin directory, for a good reason: they need to be found by Lua, not by C. NeoVim however wants to link against lpeg in advance and not dynamically, since it wants to load that module, before loading any Lua code. That's why during compile time, everything looks fine, but during runtime, it crashes: `gcc` correctly links against the symbol `luaopen_lpeg`, but the dynamic linker from Windows fails to do so, since during runtime, it can't find the dll.

From here I see 3 solutions:
1. Copy `lpeg.dll` directly into the bin folder during compilation (both in `/<distro>/bin` and `src/build<distro>/bin`. This might create conflicts between packages in the future, but it is a shortterm fix if you want to install nvim 0.10.0. If the `lpeg` package.
2. Rewrite NeoVim via a patch to ask Lua to load lpeg, instead of the dynamic linker during startup. It's a more dirty solution, but it wouldn't create any more conflicts and it would just work™
3. There is something called `rpath`, which tells the linker to add a library search path during runtime. This would be a promising aproach, especially for maintainablility, because it doesn't need to change the source, but the CMake script. Ideally we would add the lib directory relatively to the binary. However I tried to convince CMake to include the rpath, without success. Even after instructing CMake to include the linker flag directly via `-Wl,-rpath,${LPEG_LIBRARY_DIR}`, the library did not link. This indicates that either I (hopefully) did something wrong, or Windows does not support `rpath`.

I believe this is the only issue for updating to `0.10.0`, since, at least for me, neovim `0.10.0` does start up just fine, if I apply (1) manually.

**Comment by sewbacca** [comment]
FYI I implemented my proposed solution (2) in #21522 and tested it locally, with no issues.

**Comment by sewbacca** [comment]
@hjm7 The update to 0.10.1 has been merged in #21522 and is installable via `pacman`. Consider resolving this issue?
